internal/provider/resource_private_key.go

```diff
@@ -290,3 +268,88 @@ func (r *privateKeyResource) Delete(ctx context.Context, _ tfsdk.DeleteResourceR
 	// NO-OP: Returning no error is enough for the framework to remove the resource from state.
 	tflog.Debug(ctx, "Removing private key from state")
 }
+
+func (r *privateKeyResource) UpgradeState(ctx context.Context) map[int64]tfsdk.ResourceStateUpgrader {
+	schema, _ := (&privateKeyResourceType{}).GetSchema(ctx)
+	return map[int64]tfsdk.ResourceStateUpgrader{
+		// Upgrading schema v0 -> v1 will add:
+		// * `private_key_openssh` (introduced in v3.2.0)
+		// * `public_key_fingerprint_sha256` (introduced in v3.2.0)
+		// * `private_key_pkcs8`   (introduced in v4.0.0)
+		0: {
+			PriorSchema: &schema,
```

Comment:
`PriorSchema` appears to be using a schema with `Version: 1`. Should the version for the `PriorSchema` match the map key (i.e., `0`)?

Reply:
Great catch.
I have left a long explanation in the code now:
```
// NOTE: why are we using a Schema v1 to unmarshal configuration with Schema v0?
// This is possible because the way the unmarshalling works, is that if a field
// is found in the RawSchema, is then set on the State by looking for a compatible
// field in the Schema.
//
// In other words, fields that are not present in the RawSchema, are simply ignored:
// the equivalency of RawSchema and Schema is not bidirectional.
//
// This works fine _as long as_ there are no renaming or retractions from the Schema.
// If/when that happens, and a new Schema version is released, then a dedicated
// schema will be necessary.
```